lib/puppet/type/file/content_uri.rb

```diff
@@ -0,0 +1,42 @@
+require 'net/http'
+require 'uri'
+
+module Puppet
+  Puppet::Type.type(:file).newparam(:content_uri) do
+
+    desc <<-'EOT'
+      A unique identifier for file content, which must be a URI which points
+      to a file in a module. This parameter allows the content of a file resource
+      to remain consistent regardless of code updates on the master.
+
+      The normal form of a `puppet:` URI is:
+
+      `puppet:///modules/<MODULE NAME>/files/<FILE PATH>`
+
+      Note that content_uri contains "files" in its path, unlike the source parameter.
+    EOT
+
+    validate do |path|
+      begin
+        uri = URI.parse(URI.escape(path))
+      rescue
+        self.fail Puppet::Error, "Could not understand URI #{path}: #{detail}", detail
+      end
+
+      self.fail "Cannot use opaque URLs '#{source}'" unless uri.hierarchical?
+      unless uri.scheme == "puppet"
+        self.fail "Must use URLs of type puppet as content URI"
+      end
+    end
+
+    def get_from_puppet_source(source_or_content, &block)
+      request = Puppet::Indirector::Request.new(:code_content, :find, resource[:content_uri], nil, :environment => resource.catalog.environment_instance)
```

Comment:
Please delete me out of this 

Sent from Outlook Mobile




On Thu, Jan 21, 2016 at 4:49 PM -0800, "Josh Cooper" <notifications@github.com> wrote:












In lib/puppet/type/file/content_uri.rb:
> +
> +    validate do |path|
> +      begin
> +        uri = URI.parse(URI.escape(path))
> +      rescue
> +        self.fail Puppet::Error, "Could not understand URI #{path}: #{detail}", detail
> +      end
> +
> +      self.fail "Cannot use opaque URLs '#{source}'" unless uri.hierarchical?
> +      unless uri.scheme == "puppet"
> +        self.fail "Must use URLs of type puppet as content URI"
> +      end
> +    end
> +
> +    def get_from_puppet_source(source_or_content, &block)
> +      request = Puppet::Indirector::Request.new(:code_content, :find, resource[:content_uri], nil, :environment => resource.catalog.environment_instance)



@HAIL9000 I think I misread the ticket you're working on. I was originally thinking of this as being a 1 pt, add the parameter to the file type, as you suggested in hipchat. We can add the method for making the network request in https://tickets.puppetlabs.com/browse/PUP-5114, along with adding vcr related specs and API docs. So how about if we drop the get_from_puppet_source method from this PR and pick it up with the later ticket?



—
Reply to this email directly or view it on GitHub.